### ISSUE TITLE
Consider migrating tests from AppVeyor/Travis CI to Azure Pipelines

### ISSUE THREAD

**Issue by hcho3** [dmlc/xgboost#4498]
Consider migrating all tests from AppVeyor/Travis CI to [Azure Pipelines](https://azure.microsoft.com/en-us/services/devops/pipelines/).

Why migrate tests?
* OSX tests on Travis CI are quite slow and have proven flaky from time to time (e.g. #4487,  #4448).
* AppVeyor allows only a single job to run at a time; Azure Pipelines allows for 10 parallel jobs
* Azure Pipelines can run both Windows and Mac OSX tests (note: no GPUs.)
* Azure Pipelines is provided free of charge to open-source projects

TODO. Test Azure Pipelines for performance and stability

cc @sriramch

**Comment by CodingCat** [comment]
+10000 on getting rid of AppVeyor

**Comment by hcho3** [comment]
I played with Azure Pipelines a bit, and so far it's been pleasure to work with.
* Builds start almost instantaneously and run fast.
* In a single YAML file, I can use Linux, Mac OSX, and Windows agents.
* Jobs can run in parallel, but some jobs can be made dependent on other jobs.

See example: [azure-pipelines.yml](https://github.com/dmlc/treelite/blob/600afd55d1fa9bb94fc88fd3a3043cb2d5b20651/azure-pipelines.yml), [build result](https://dev.azure.com/hcho3/treelite/_build/results?buildId=49)

**Comment by trivialfis** [comment]
Can't wait.

**Comment by trams** [comment]
I am also interested in seeing Azure Pipelines instead of AppVeyor

**Comment by hcho3** [comment]
We've set up GitHub Actions (#5917) and moved some tests from AppVeyor to GitHub Actions (#5922). There is currently a nondeterministic test that's blocking a further migration away from AppVeyor. 

We are keeping Travis CI because we need access to MacOS High Sierra to build XGBoost Python wheel. (Expectation is that it will be compatible with last three major releases, High Sierra, Mojave, and Catalina.) GitHub Actions only provides MacOS Catalina.

Closing this in favor of #5891.

**Comment by trivialfis** [comment]
@hcho3 I don't think we are running 3 versions of osx on travis right?

**Comment by trivialfis** [comment]
And another thing is, our test on Travis now take 1.5+ hour to run for each PR.  And multiple PRs cannot be run in parallel.

**Comment by hcho3** [comment]
> And multiple PRs cannot be run in parallel.

Travis CI runs multiple Linux jobs just fine. It's just that MacOS agent pool is really small, so our jobs get stuck in the global queue for a long time.

**Comment by hcho3** [comment]
> I don't think we are running 3 versions of osx on travis right?

No. We assume backward compatibility: if we can build and run XGBoost on MacOS High Sierra, then the generated binary should also be compatible with MacOS Mojave/Catalina.

**Comment by trivialfis** [comment]
Got it.  Thanks for the explanation.

**Comment by hcho3** [comment]
Reopening, since Travis CI seems to be quite slow nowadays. The tests remain in queue for 2-3 hours. (Once they start, they are usually fairly quick.)

A proposal: Migrate all MacOS tests to GitHub Actions. Travis CI should only run on the master branch, to generate and upload nightly builds (Python wheels). Hopefully, discrepancy between MacOS releases isn't too big so that code that works in one release would break in another too often.  @trivialfis WDYT?

**Comment by trivialfis** [comment]
@hcho3 I'm not sure about the ecosystem on osx.  I will leave it to others to decide. ;-)  @CodingCat seems also fond of osx.

**Comment by CodingCat** [comment]
I think the backward compatibility of osx is generally guaranteed....

**Comment by hcho3** [comment]
Travis CI seems fine now.
